omniscidb/Shared/quantile.h

```diff
@@ -0,0 +1,343 @@
+    }
+
+    bool operator!=(const Iterator& other) const { return !(*this == other); }
+
+    bool operator<(const Iterator& other) const {
+      if (chunk_idx_ != other.chunk_idx_) {
+        return chunk_idx_ < other.chunk_idx_;
+      }
+      return chunk_offs_ < other.chunk_offs_;
+    }
+
+    bool operator<=(const Iterator& other) const {
+      if (chunk_idx_ != other.chunk_idx_) {
+        return chunk_idx_ < other.chunk_idx_;
+      }
+      return chunk_offs_ <= other.chunk_offs_;
+    }
+
+    T& operator*() const {
+      auto data = reinterpret_cast<T*>((*chunks_)[chunk_idx_].data);
+      return data[chunk_offs_];
+    }
+
+   private:
+    const std::vector<Chunk>* chunks_;
+    // Current chunk index. Can be equal to size of chunks_ vector for `end` iterator.
+    size_t chunk_idx_;
+    // Offset in the current chunk. Should always be less than chunk size when the
+    // index points to a valid chunk.
+    size_t chunk_offs_;
+  };
+
+  ChunkedArray(SimpleAllocator* allocator) : allocator_(allocator), cur_idx_(0) {}
+
+  template <typename T>
+  void push(T value) {
+    // Check if we need to allocate a new chunk.
+    if (chunks_.empty() || cur_idx_ == chunks_.back().size) {
+      // Allocator is most probably a RowSetMemoryOwner object. It is not supposed to be
+      // used to allocate very small objects, so we start with 1 KB and double it each
+      // time with 64KB limit.
+      size_t size_to_allocate = std::max((size_t)64, (size_t)1 << chunks_.size()) << 10;
+      Chunk chunk{allocator_->allocate(size_to_allocate), size_to_allocate / sizeof(T)};
+      chunks_.emplace_back(chunk);
+      cur_idx_ = 0;
+    }
+    auto data = reinterpret_cast<T*>(chunks_.back().data);
+    data[cur_idx_++] = value;
+  }
+
+  void merge(const ChunkedArray& other) {
+    if (!other.chunks_.empty()) {
+      // We are not going to add values to the current last chunk anymore, so fix-up
+      // its size for proper iteration.
+      if (!chunks_.empty()) {
+        chunks_.back().size = cur_idx_;
+      }
+      chunks_.insert(chunks_.end(), other.chunks_.begin(), other.chunks_.end());
+      cur_idx_ = other.cur_idx_;
+    }
+  }
+
+  template <typename T>
+  Iterator<T> begin() {
+    return Iterator<T>(&chunks_, 0, 0);
+  }
+
+  template <typename T>
+  Iterator<T> end() {
+    // Chunk offset in iterator is always expected to be less than chunk size. So, end
+    // iterator for a full chunk is supposed to point the start of the next chunk.
+    // Take care of it if all chunks are full.
+    if (!chunks_.empty() && cur_idx_ == chunks_.back().size) {
+      return Iterator<T>(&chunks_, chunks_.size(), 0);
+    }
+    return Iterator<T>(&chunks_, chunks_.size() - 1, cur_idx_);
+  }
+
+  bool empty() const { return chunks_.empty(); }
+
+  size_t size() const {
+    if (chunks_.empty()) {
+      return 0;
+    }
+
+    size_t res = cur_idx_;
+    for (size_t i = 0; i < chunks_.size() - 1; ++i) {
+      res += chunks_[i].size;
```

Comment:
The size does not match the number of pushed elements here, right? Looks like vector's capacity.